### PR TITLE
Try some basic unit tests

### PR DESCRIPTION
**Pull request by capnrefsmmat**
WIP. Will write some more tests Soonish to see what happens (and how well our CI works).

### CODE REVIEW

**Comment by capnrefsmmat** [comment]
Apparently there's a nice package for testing plotting, so we can also do that to avoid regressing our plotting features: https://github.com/r-lib/vdiffr

**Comment by capnrefsmmat** [comment]
Ok, I've successfully set up mockery and vdiffr to do tests of functions that use the network (like `covidcast_signal`) and tests of our plotting/mapping functions. They work great on simple examples.

I also added `DEVELOP.md` with instructions on what to run and a PR checklist.

I think it'll be easy to make a few different plot examples (by saving a few real data frames) and save those as the key plotting tests. We could also add tests for issues like #134 when they're fixed.

**Comment by capnrefsmmat** [comment]
There are a lot more tests we should write, but I likely won't have time to write more for a while. So it's probably best if we merge this now, and then file an issue with a list of things we should add tests for. Then the testing framework will be in place for everyone to use for their new PRs, and we can fill in the testing gaps over the next few weeks.

**Comment by ryantibs** [comment]
This looks great, thanks for starting this off!. I won't be able to review this in any level of detail in the next few days, too bogged down with other things. Are any of @nmdefries @JedGrabman @sarah-colq @sgsmob @undefx @sangwon-hyun @nloliveira @capolitsch @brookslogan available?

**Comment by sgsmob** [comment]
I can't assign myself to review this but I will

File: R-packages/covidcast/DESCRIPTION [hunk 1]
```diff
@@ -52,5 +52,9 @@ RoxygenNote: 7.1.1
 Suggests: 
     knitr,
     rmarkdown,
-    gridExtra
+    gridExtra,
+    testthat,
```
Comment (sgsmob) [inline R-packages/covidcast/DESCRIPTION:56]: Please alphabetize these lists

File: R-packages/covidcast/tests/testthat/test-cor.R [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+  expect_equal(covidcast_cor(foo, bar, by = "geo_value"),
+               expected)
+
+  # Try permuting order
+  expect_equal(covidcast_cor(foo, bar[c(2, 1, 4, 3), ],
+                             by = "geo_value"),
+               expected)
+})
+
+test_that("basic lagged correlations", {
```
Comment (sgsmob) [inline R-packages/covidcast/tests/testthat/test-cor.R:33]: Please add a comment explaining the purpose of this test.

File: R-packages/covidcast/tests/testthat/test-cor.R [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    geo_value = "pa",
+    value = 1
+  )
+  expect_equal(covidcast_cor(foo, bar, dt_x = 1),
+               expected)
+  expect_equal(covidcast_cor(foo, bar, dt_y = 1),
+               expected)
+})
+
+test_that("lags are in correct direction", {
```
Comment (sgsmob) [inline R-packages/covidcast/tests/testthat/test-cor.R:60]: Please add a comment explaining the purpose of this test.

File: R-packages/covidcast/tests/testthat/test-covidcast.R [hunk 4]
```diff
@@ -0,0 +1,26 @@
+library(covidcast)
+library(mockery)
+
+test_that("covidcast_meta", {
```
Comment (sgsmob) [inline R-packages/covidcast/tests/testthat/test-covidcast.R:4]: Please add a comment explaining this test

File: R-packages/covidcast/tests/testthat/test-covidcast.R [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+                   max_time = as.Date(c("2020-01-02", "2020-10-03")),
+                   max_issue = as.Date(c("2020-04-04", "2020-11-01"))
+                 ),
+                 class = c("covidcast_meta", "data.frame")
+               ))
+
+  stub(covidcast_meta, ".request",
+       list(message = "argle-bargle"))
+
+  expect_error(covidcast_meta())
```
Comment (sgsmob) [inline R-packages/covidcast/tests/testthat/test-covidcast.R:25]: Does this allow specification of which error is expected?

File: R-packages/covidcast/tests/testthat/test-plot.R [hunk 6]
```diff
@@ -0,0 +1,81 @@
+library(testthat)
+library(vdiffr)
+library(dplyr)
+
+# Contexts are no longer required or recommended in testthat, but vdiffr still
+# wants one to place the figure files correctly. See
+# https://github.com/r-lib/vdiffr/issues/71
+context("plot")
+
+test_that("simple line graph", {
```
Comment (sgsmob) [inline R-packages/covidcast/tests/testthat/test-plot.R:10]: Please leave a comment describing the purpose of this test

File: R-packages/covidcast/tests/testthat/test-plot.R [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+    line_params = list(
+      xlab = "Day",
+      ylab = "Penguinocity",
+      stderr_bands = TRUE,
+      stderr_alpha = 0.3
+    )
+  ))
+})
+
+test_that("state line graphs", {
```
Comment (sgsmob) [inline R-packages/covidcast/tests/testthat/test-plot.R:35]: Please leave a comment describing the purpose of this test

File: R-packages/covidcast/tests/testthat/test-plot.R [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+                      plot(filter(fb_state, geo_value %in% c("pa", "tx", "ny")),
+                           plot_type = "line",
+                           line_params = list(stderr_bands = TRUE)))
+
+  expect_doppelganger("state line graph with range",
+                      plot(fb_state, plot_type = "line",
+                           range = c(0, 10)))
+})
+
+test_that("simple state choropleths", {
```
Comment (sgsmob) [inline R-packages/covidcast/tests/testthat/test-plot.R:51]: Please leave a comment describing the purpose of this test

File: R-packages/covidcast/tests/testthat/test-wrangle.R [hunk 9]
```diff
@@ -0,0 +1,82 @@
+library(covidcast)
+library(dplyr)
+
+test_that("aggregate and shift", {
```
Comment (sgsmob) [inline R-packages/covidcast/tests/testthat/test-wrangle.R:4]: Please leave comments describing the purpose of the tests in this file

File: R-packages/covidcast/tests/testthat/test-wrangle.R [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+               structure(data.frame(
+                 data_source = c(rep("bar", 3), rep("foo", 3)),
+                 signal = c(rep("bar", 3), rep("foo", 3)),
+                 geo_value = rep(c("pa", "ri", "tx"), 2),
+                 time_value = as.Date("2020-01-01"),
+                 dt = 0,
+                 value = c(4, 6, 5, 1, 3, 2)),
+                 class = c("covidcast_signal_long", "data.frame")))
+
+  # Now try it long in the first place. Currently fails because wide format does
```
Comment (sgsmob) [inline R-packages/covidcast/tests/testthat/test-wrangle.R:78]: If these tests are not working, they should just be removed from this file entirely.

**Review by sgsmob — CHANGES_REQUESTED** [review]
This will be a great asset to this repo!

Not entirely sure how to validate an initial commit of the charts but I think they are probably ok.  Will approve once some more comments are added.

File: R-packages/covidcast/tests/testthat/test-covidcast.R (hunk 5 above)
Comment (capnrefsmmat) [reply R-packages/covidcast/tests/testthat/test-covidcast.R]: Not if `covidcast_meta()` uses `stop()`, which does not specify a specific error type -- unless you regexp-match against the error message.

I've just changed `covidcast_meta()` to use `abort()` with a specific error class, so we can catch and handle it, and the test can verify it was raised correctly.

**Comment by capnrefsmmat** [comment]
Pushed fixes. Rather than adding a bunch of comments to the tests, I tried to change their names so that when you read them, they're meaningful, such as

```r
test_that("covidcast_meta raises error when API signals one", {
```

**Comment by capnrefsmmat** [comment]
Also, regarding the commented-out test code in `test-wrangle.R`: I can use `skip()` to skip that instead of commenting out, but that also prints a message every time you run the tests. The issue is that I want to change the code to make that test pass, but it doesn't yet; would you prefer the code be removed entirely and returned when the issues is fixed, leave it commented out, or have it skipped and noisily complain to encourage people to fix the wrangling code?

**Comment by sgsmob** [comment]
> Also, regarding the commented-out test code in `test-wrangle.R`: I can use `skip()` to skip that instead of commenting out, but that also prints a message every time you run the tests. The issue is that I want to change the code to make that test pass, but it doesn't yet; would you prefer the code be removed entirely and returned when the issues is fixed, leave it commented out, or have it skipped and noisily complain to encourage people to fix the wrangling code?

A TODO in the comments at the top would be good.

**Review by sgsmob — APPROVED** [review]

**Pull request merged**